Cross compile to Scala Native

Duhemm opened the pull request:
The next release of Scala Native is coming soon, and it will add support for test frameworks using sbt's test interface (scala-native/scala-native#755). While implementing the test interface, I cross compiled utest to Scala Native.

Here are the changes that I did to get it to cross compile and work with Scala Native. Obviously, Scala Native 0.3.0 is not released yet, and you'll need to build Scala Native locally to test it. Here's what it looks like: https://travis-ci.org/Duhemm/scala-native/jobs/242582498#L2665-L2697

## How to try it

 - Here's the branch that adds the test interface to Scala Native: https://github.com/Duhemm/scala-native/tree/topic/test-interface (use that instead of Scala Native's `master` branch if you happen to read that before the aforementioned PR is merged).
 - Here's how to get started with Scala Native: http://www.scala-native.org/en/latest/user/setup.html

After your system is set up, you can build Scala Native locally:

```
$ git clone https://github.com/scala-native/scala-native.git
$ cd scala-native
$ sbt rebuild
```

Then, build and `publishLocal`ly this PR. You can use the published artifact in a Scala Native project [as shown here](https://github.com/Duhemm/scala-native/blob/8939e764fff8faae8a77df9b872df1414dc403c1/build.sbt#L431-L432). Write your tests as usual and run them using `test` in sbt.

It would be great if we could get this PR in soon after Scala Native is released (I'll update it with the correct version numbers). Do you see things that need to be addressed in this PR while we wait for the next release of Scala Native?

xuwei-k commented:
FYI, I found some bugs.
- https://github.com/scala-native/scala-native/issues/842
- https://github.com/rickynils/scalacheck/pull/337#issuecomment-308925196

japgolly commented:
Thanks @xuwei-k

Duhemm commented:
Thanks for the report @xuwei-k! These bugs will be fixed in Scala Native 0.3.1. And thank you for your fix 😉 

@japgolly I've updated the PR with the changes you requested. Note that I also needed to modify some of the tests:

 - Two tests were using `x / 0` to throw an exception. Unfortunately, in Scala Native, this results in a crash rather than just an exception thrown. I changed these tests to use `Iterator.empty.next()` instead. ([here for instance](https://github.com/lihaoyi/utest/pull/114/files#diff-b9acac2b9c808d83b98b229d59b9f1fcR181))
 - The asynchronous tests try to schedule the tests so that they are completed after 2 seconds. This is not supported in Scala Native at the moment, as we can complete computations either immediately or at the end of the program (the second option wouldn't work for utest, as it will consider that the test never completed). I simply bypass the "not-completed-yet" check when we're running with Scala Native. ([here for instance](https://github.com/lihaoyi/utest/pull/114/files#diff-7d214138ad20b9f2138be3467f6e4318R23))
 - Two tests are using `0.asInstanceOf[String]` to throw a `ClassCastException`, but Scala native doesn't throw an exception in this case. This is a bug in Scala Native that has been filled here: scala-native/scala-native#858. ([here for instance](https://github.com/lihaoyi/utest/pull/114/files#diff-7d214138ad20b9f2138be3467f6e4318R62))

Duhemm commented:
Could someone restart the build, please?

lihaoyi commented:
Just kicked it

lihaoyi commented:
After several more kicks, it has turned green

lihaoyi commented:
@japgolly up to you to merge/publish this, if you so desire

japgolly commented:
Mergerino!

japgolly commented:
Maybe I was too care-free in merging this. I just gave it a try locally. It compiles fine but when I run it:
```
[info] Linking (982 ms)
[info] Discovered 1893 classes and 11570 methods
[info] Optimizing (1277 ms)
[info] Generating intermediate code (264 ms)
[info] Produced 58 files
[info] Compiling to native code (1159 ms)
[error] /usr/bin/ld: /home/golly/projects/utest/utest/native/target/scala-2.11/native/lib/platform.c.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
[error] /usr/bin/ld: /home/golly/projects/utest/utest/native/target/scala-2.11/native/lib/safepoint.c.o: relocation R_X86_64_32 against symbol `scalanative_safepoint_trigger' can not be used when making a shared object; recompile with -fPIC
[error] /usr/bin/ld: /home/golly/projects/utest/utest/native/target/scala-2.11/native/lib/unistd.c.o: relocation R_X86_64_PC32 against symbol `symlink@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
[error] /usr/bin/ld: final link failed: Bad value
[error] clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
[info] Linking native code (132 ms)
[info] Starting process '/home/golly/projects/utest/utest/native/target/scala-2.11/utest-out' on port '9862'.
Exception in thread "Thread-1178" java.io.IOException: Cannot run program "/home/golly/projects/utest/utest/native/target/scala-2.11/utest-out": error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at sbt.SimpleProcessBuilder.run(ProcessImpl.scala:349)
	at sbt.AbstractProcessBuilder.run(ProcessImpl.scala:128)
	at sbt.AbstractProcessBuilder$$anonfun$runBuffered$1.apply(ProcessImpl.scala:159)
	at sbt.AbstractProcessBuilder$$anonfun$runBuffered$1.apply(ProcessImpl.scala:159)
	at sbt.BufferedLogger.buffer(BufferedLogger.scala:25)
	at sbt.AbstractProcessBuilder.runBuffered(ProcessImpl.scala:159)
	at sbt.AbstractProcessBuilder.$bang(ProcessImpl.scala:156)
	at scala.scalanative.sbtplugin.testinterface.ComRunner$$anon$1.run(ComRunner.scala:35)
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 8 more
[trace] Stack trace suppressed: run last utestNative/nativetest:executeTests for the full output.
[error] (utestNative/nativetest:executeTests) java.lang.Exception: Couldn't communicate with remote runner.
[error] Total time: 5 s, completed 26/06/2017 10:00:11 AM
```

I don't really have time for this :(
Any suggestions, O' thine of more experience than I?

densh commented:
@japgolly Can you please file an issue on Scala Native with steps to reproduce and with the details on the environment? This looks like either an environment misconfiguration or a bug with should be able to fix in `0.3.1` which about to come this week.

ShaneDelmore commented:
Sorry to hear this isn't flawless but for some positive feedback for the work you are all doing, I used it this weekend and it was great, made my day to be able to write a command line application with unit tests and then see the app run and do real work in 0.035s 👍

japgolly commented:
Raised https://github.com/scala-native/scala-native/issues/881

And for what it's worth, I think all the work being done in this space is fucking awesome! I harbour much appreciation to all involved. Love seeing the progress :+1:
